src/ConnectedMode.UnitTests/ServerSentEvents/QualityProfile/QualityProfileServerEventsListenerTests.cs

```diff
@@ -0,0 +1,81 @@
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using SonarLint.VisualStudio.ConnectedMode.QualityProfiles;
+using SonarLint.VisualStudio.ConnectedMode.ServerSentEvents.QualityProfile;
+using SonarLint.VisualStudio.TestInfrastructure;
+using SonarQube.Client.Models.ServerSentEvents.ClientContract;
+
+namespace SonarLint.VisualStudio.ConnectedMode.UnitTests.ServerSentEvents.QualityProfile;
+
+[TestClass]
+public class QualityProfileServerEventsListenerTests
+{
+    [TestMethod]
+    public void MefCtor_CheckExports()
+    {
+        MefTestHelpers.CheckTypeCanBeImported<QualityProfileServerEventsListener, IQualityProfileServerEventsListener>(
+            MefTestHelpers.CreateExport<IQualityProfileServerEventSource>(),
+            MefTestHelpers.CreateExport<IQualityProfileUpdater>());
+    }
+
+    [TestMethod]
+    public void MefCtor_CheckIsSingleton()
+    {
+        MefTestHelpers.CheckIsSingletonMefComponent<QualityProfileServerEventsListener>();
+    }
+
+    [TestMethod]
+    public void ListenAsync_CallsUpdaterForEachEventUntilNullEvent()
+    {
+        var events = new[]
+        {
+            Mock.Of<IQualityProfileEvent>(),
+            Mock.Of<IQualityProfileEvent>(),
+            Mock.Of<IQualityProfileEvent>(),
+            null
+        };
+
+        var tcs = Enumerable.Range(0, events.Length).Select(_ => new TaskCompletionSource<IQualityProfileEvent>()).ToArray();
+
+        var eventSourceMock = new Mock<IQualityProfileServerEventSource>();
+        eventSourceMock.Setup(x => x.GetNextEventOrNullAsync()).Returns(new Queue<Task<IQualityProfileEvent>>(tcs.Select(x => x.Task)).Dequeue);
+        
+        var updaterMock = new Mock<IQualityProfileUpdater>();
+
+        var testSubject = new QualityProfileServerEventsListener(eventSourceMock.Object, updaterMock.Object);
+        var task = testSubject.ListenAsync();
+
+        for (var index = 0; index < tcs.Length; index++)
+        {
+            var taskCompletionSource = tcs[index];
+            taskCompletionSource.SetResult(events[index]);
+            task.IsCompleted.Should().Be(index == tcs.Length - 1);
+        }
```

Comment:
Reading this test made my brain hurt so I debugged through it so I could understand it better, and the debugging experience was weird too.

It feels like there are more data structures than are really needed, and that there isn't a clean separation between the setup code and the execution (i.e. setting the `taskCompletionSource` result in the `for` loop after execution has started).

I had a go at writing it in an alternative way and came up with the following:

```
    [TestMethod]
    public async Task ListenAsync_CallsUpdaterForEachEventUntilNullEvent()
    {
        var currentEventIndex = 0;
        var eventTasks = new Task<IQualityProfileEvent>[]
        {
            Task.FromResult(Mock.Of<IQualityProfileEvent>()),
            Task.FromResult(Mock.Of<IQualityProfileEvent>()),
            Task.FromResult(Mock.Of<IQualityProfileEvent>()),
            Task.FromResult<IQualityProfileEvent>(null),
        };

        var eventSourceMock = new Mock<IQualityProfileServerEventSource>();
        eventSourceMock.Setup(x => x.GetNextEventOrNullAsync()).Returns(() => eventTasks[currentEventIndex++]);

        var updaterMock = new Mock<IQualityProfileUpdater>();

        var testSubject = new QualityProfileServerEventsListener(eventSourceMock.Object, updaterMock.Object);
        await testSubject.ListenAsync();

        eventSourceMock.Verify(x => x.GetNextEventOrNullAsync(), Times.Exactly(eventTasks.Length));
        updaterMock.Verify(x => x.UpdateAsync(), Times.Exactly(eventTasks.Length - 1));
    }
```

This version doesn't have an equivalent of the check `task.IsCompleted.Should().Be(...)`,
However, I don't think it needs it: checking the correct number of server and updater calls should be enough.

What do you think?

Reply:
The point of checking IsCompleted was to indicate/verify that it's a long running action. If you think that it's not worth making the test more complicated to test that, then your test is better. I'm personally not sure if it's important. But tbf, I also not fully understand why my test is confusing